### 1. Classification
    분류 = 정해진 종류 중 하나를 예측하는 문제

    예를 들면,
        iris          -> setosa / versicolor / virginica
        breast cancer -> malignant / benign

### 2. Iris 데이터셋 실습

#### 1. 데이터 불러오고 구조 확인

In [47]:
from sklearn.datasets import load_iris

iris = load_iris()

X = iris.data   # feature
y = iris.target

print("X shape:", X.shape)
print("y shape:", y.shape)
print()

print("특성 이름:", iris.feature_names)
print("클래스 이름:", iris.target_names)
print()

print("X 앞 5개:")
print(X[:5])
print()

print("y 앞 5개:")
print(y[:5])

X shape: (150, 4)
y shape: (150,)

특성 이름: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
클래스 이름: ['setosa' 'versicolor' 'virginica']

X 앞 5개:
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]

y 앞 5개:
[0 0 0 0 0]


X shape: (150, 4)
- 데이터 개수 = 150개
- 특성 개수 = 4개
- 정답 라벨 개수 = 150개

- 즉, X는 아래처럼 생긴 데이터이다.
    - 꽃받침 길이
    - 꽃받침 너비
    - 꽃잎 길이
    - 꽃잎 너비

y shape: (150,)
- y는 각 꽃이 어떤 품종인지 나타낸다.
    - 0 = setosa
    - 1 = versicolor
    - 2 = virginica

iris는 클래스가 3개이므로 다중 분류 문제다.

# 3. iris 분류모델 만들기

In [48]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

# 1. 데이터 불러오기
iris = load_iris()

X = iris.data
y = iris.target

# 2. train/test 분리
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size = 0.2,
    random_state = 42
)

# 3. 모델 생성
model = KNeighborsClassifier(n_neighbors=3)

# 4. 학습
model.fit(X_train, y_train)

# 5. 예측
pred = model.predict(X_test)

# 6.평가
accuracy = model.score(X_test, y_test)

print("예측값:", pred)
print("실제값:", y_test)
print("정확도:", accuracy)

예측값: [1 0 2 1 1 0 1 2 1 1 2 0 0 0 0 1 2 1 1 2 0 2 0 2 2 2 2 2 0 0]
실제값: [1 0 2 1 1 0 1 2 1 1 2 0 0 0 0 1 2 1 1 2 0 2 0 2 2 2 2 2 0 0]
정확도: 1.0


모델이 예측한 값과 실제 정답이 같으면 맞힌거다

정확도는 전체 테스트 데이터 중 맞힌 비율이다.
- 정확도 = 맞힌 개수 / 전체 테스트 데이터 개수

# 4. 틀린 데이터 직접 확인하기
- 정확도만 보면 감이 약하니까, 어떤 데이터를 틀렸는지 직접 확인

In [49]:
wrong_index = pred != y_test    # 예측값과 실제값이 다른 위치를 찾는 부분

print("틀린 예측값:", pred[wrong_index])
print("틀린 실제값:", y_test[wrong_index])
print("틀린 데이터 개수:", wrong_index.sum())

틀린 예측값: []
틀린 실제값: []
틀린 데이터 개수: 0


# 5. 클래스 이름으로 출력하기
- 숫자 0, 1, 2만 보면 직관이 떨어지니까 클래스 이름으로 바꾸면

In [50]:
print("예측 클래스 이름:", iris.target_names[pred])
print()
print("실제 클래스 이름:", iris.target_names[y_test])

예측 클래스 이름: ['versicolor' 'setosa' 'virginica' 'versicolor' 'versicolor' 'setosa'
 'versicolor' 'virginica' 'versicolor' 'versicolor' 'virginica' 'setosa'
 'setosa' 'setosa' 'setosa' 'versicolor' 'virginica' 'versicolor'
 'versicolor' 'virginica' 'setosa' 'virginica' 'setosa' 'virginica'
 'virginica' 'virginica' 'virginica' 'virginica' 'setosa' 'setosa']

실제 클래스 이름: ['versicolor' 'setosa' 'virginica' 'versicolor' 'versicolor' 'setosa'
 'versicolor' 'virginica' 'versicolor' 'versicolor' 'virginica' 'setosa'
 'setosa' 'setosa' 'setosa' 'versicolor' 'virginica' 'versicolor'
 'versicolor' 'virginica' 'setosa' 'virginica' 'setosa' 'virginica'
 'virginica' 'virginica' 'virginica' 'virginica' 'setosa' 'setosa']


# 6. Breast Cancer 데이터셋
- Breast cancer 데이터셋은 종양이 악성인지 양성인지 분류하는 데이터다.
    - malignant = 악성
    - benign = 양성

In [51]:
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()

X = cancer.data
y = cancer.target

print("X shape:", X.shape)
print("y_shape:", y.shape)
print()

print("특성 이름 개수:", len(cancer.feature_names))
print("클래스 이름:", cancer.target_names)
print()

print("X 앞 2개:")
print(X[:2])
print()

print("y 앞 10개:")
print(y[:10])

X shape: (569, 30)
y_shape: (569,)

특성 이름 개수: 30
클래스 이름: ['malignant' 'benign']

X 앞 2개:
[[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
  1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
  6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
  1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
  4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 1.326e+03 8.474e-02 7.864e-02 8.690e-02
  7.017e-02 1.812e-01 5.667e-02 5.435e-01 7.339e-01 3.398e+00 7.408e+01
  5.225e-03 1.308e-02 1.860e-02 1.340e-02 1.389e-02 3.532e-03 2.499e+01
  2.341e+01 1.588e+02 1.956e+03 1.238e-01 1.866e-01 2.416e-01 1.860e-01
  2.750e-01 8.902e-02]]

y 앞 10개:
[0 0 0 0 0 0 0 0 0 0]


출력은 대략 아래처럼 나온다

- X shape: (569, 30)
- Y shape: (569,)

의미:
- 데이터 개수 = 569개
- 입력 특성 개수 = 30개
- 정답 라벨 개수 = 569개

클래스는 2개다
- malignant
- benign

따라서 breast cancer 데이터셋은 이진 분류 문제다.

# 7. Breast Cancer 분류 모델

In [52]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline

# 1. 데이터 불러오기
cancer = load_breast_cancer()

X = cancer.data
y = cancer.target

# 2. train/test 분리
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size = 0.2,
    random_state = 42
)

# 3. 표준화 + KNN 모델
model = Pipeline([
    ("scaler", StandardScaler()),
    ("knn", KNeighborsClassifier(n_neighbors=3))
])

# 4. 학습
model.fit(X_train, y_train)

# 5. 예측
pred = model.predict(X_test)

# 6. 평가
accuracy = model.score(X_test, y_test)

print("예측값:", pred)
print("실제값:", y_test)
print("정확도:", accuracy)

예측값: [1 0 0 1 1 0 0 0 0 1 1 0 1 1 1 0 1 1 1 0 1 1 0 1 1 1 1 1 1 0 1 1 1 1 1 1 0
 1 0 1 1 0 1 1 1 1 1 1 1 1 0 0 0 1 1 1 1 0 0 1 1 0 0 1 1 1 0 0 1 1 0 0 1 0
 1 1 1 1 1 1 0 1 0 0 0 0 0 0 1 1 1 0 1 1 1 1 0 0 1 0 0 1 0 0 1 1 1 0 1 1 0
 1 1 0]
실제값: [1 0 0 1 1 0 0 0 1 1 1 0 1 0 1 0 1 1 1 0 0 1 0 1 1 1 1 1 1 0 1 1 1 1 1 1 0
 1 0 1 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 1 1 1 0 0 1 1 0 0 1 0
 1 1 1 0 1 1 0 1 0 0 0 0 0 0 1 1 1 1 1 1 1 1 0 0 1 0 0 1 0 0 1 1 1 0 1 1 0
 1 1 0]
정확도: 0.9473684210526315


# 8. 이진 분류 결과 해석

breast cancer 데이터셋에서 y는 숫자로 나온다.

- 0 = malignant
- 1 = benign

그래서 예측값이 아래처럼 나온다면
    # [1 0 1 1 0 . . .]

이 뜻이다
- 1 -> benign
- 0 -> malignant

클래스 이름으로 바꿔서 보면 더 쉽다.

In [53]:
print("예측 클래스 이름:", cancer.target_names[pred])
print()
print("실제 클래스 이름:", cancer.target_names[y_test])

예측 클래스 이름: ['benign' 'malignant' 'malignant' 'benign' 'benign' 'malignant'
 'malignant' 'malignant' 'malignant' 'benign' 'benign' 'malignant'
 'benign' 'benign' 'benign' 'malignant' 'benign' 'benign' 'benign'
 'malignant' 'benign' 'benign' 'malignant' 'benign' 'benign' 'benign'
 'benign' 'benign' 'benign' 'malignant' 'benign' 'benign' 'benign'
 'benign' 'benign' 'benign' 'malignant' 'benign' 'malignant' 'benign'
 'benign' 'malignant' 'benign' 'benign' 'benign' 'benign' 'benign'
 'benign' 'benign' 'benign' 'malignant' 'malignant' 'malignant' 'benign'
 'benign' 'benign' 'benign' 'malignant' 'malignant' 'benign' 'benign'
 'malignant' 'malignant' 'benign' 'benign' 'benign' 'malignant'
 'malignant' 'benign' 'benign' 'malignant' 'malignant' 'benign'
 'malignant' 'benign' 'benign' 'benign' 'benign' 'benign' 'benign'
 'malignant' 'benign' 'malignant' 'malignant' 'malignant' 'malignant'
 'malignant' 'malignant' 'benign' 'benign' 'benign' 'malignant' 'benign'
 'benign' 'benign' 'benign' 'maligna

### 정리
- 클래스가 2개면 이진분류
- 클래스가 3개 이상이면 다중 분류